examples/rhat_monitor.cpp

```diff
@@ -0,0 +1,365 @@
+      return false;
+    data_[write_idx] = T(std::forward<Args>(args)...);
+    write_idx_.store(next, std::memory_order_release);
+    return true;
+  }
+
+  bool pop(T &out) noexcept {
+    auto read_idx = read_idx_.load(std::memory_order_relaxed);
+    if (read_idx == write_idx_.load(std::memory_order_acquire))
+      return false;
+    out = std::move(data_[read_idx]);
+    auto next = read_idx + 1;
+    if (next == Capacity)
+      next = 0;
+    read_idx_.store(next, std::memory_order_release);
+    return true;
+  }
+
+  std::size_t capacity() const noexcept { return Capacity; }
+
+private:
+  std::vector<T> data_;
+  alignas(std::hardware_destructive_interference_size)
+      std::atomic<std::size_t> read_idx_{0};
+  alignas(std::hardware_destructive_interference_size)
+      std::atomic<std::size_t> write_idx_{0};
+};
+
+constexpr std::size_t RING_CAPACITY = 64;
+
+using Queue = RingBuffer<SampleStats, RING_CAPACITY>;
+
+static void write_csv_header(std::ofstream &out, std::size_t dim) {
+  out << "chain,iteration,log_density";
+  for (std::size_t d = 0; d < dim; ++d)
+    out << ",theta[" << d << "]";
+  out << '\n';
+}
+
+static void write_csv(const std::string &path, std::size_t dim,
+                      const std::vector<Sample> &samples) {
+  std::ofstream out(path, std::ios::binary); // binary for Windows consistency
+  if (!out)
+    throw std::runtime_error("could not open file: " + path);
+  write_csv_header(out, dim);
+  for (const auto &sample : samples)
+    sample.write_csv(out, dim);
+}
+
+class WelfordAccumulator {
+public:
+  WelfordAccumulator() : n_(0), mean_(0.0), M2_(0.0) {}
+
+  void push(double x) {
+    ++n_;
+    const double delta = x - mean_;
+    mean_ += delta / static_cast<double>(n_);
+    const double delta2 = x - mean_;
+    M2_ += delta * delta2;
+  }
+
+  std::size_t count() const { return n_; }
+
+  double mean() const { return mean_; }
+
+  double sample_variance() const {
+    return n_ > 1 ? (M2_ / static_cast<double>(n_ - 1))
+                  : std::numeric_limits<double>::quiet_NaN();
+  }
+
+  void reset() {
+    n_ = 0;
+    mean_ = 0.0;
+    M2_ = 0.0;
+  }
+
+private:
+  std::size_t n_;
+  double mean_;
+  double M2_;
+};
+
+template <class Sampler> class ChainTask {
+public:
+  ChainTask(std::size_t chain_id, std::size_t draws_per_chain, Sampler &sampler,
+            Queue &q, std::latch &start_gate)
+      : chain_id_(chain_id), draws_per_chain_(draws_per_chain),
+        sampler_(sampler), sample_(chain_id, sampler.dim(), draws_per_chain),
+        q_(q), start_gate_(start_gate) {}
+
+  void operator()(std::stop_token st) {
+    initiated_qos();
+    start_gate_.arrive_and_wait();
+    for (std::size_t iter = 0; iter < draws_per_chain_; ++iter) {
+      auto [logp, theta] = sampler_();
+      logp_stats_.push(logp);
+      sample_.append_draw(logp, std::move(theta));
+      q_.emplace(logp_stats_.count(), logp_stats_.mean(),
+                 logp_stats_.sample_variance());
```

Comment:
It feels like this could be simpler if you had a function on the welford accumulator that returned a SampleStats